File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/Chan.java
/*
 * KurobaEx - *chan browser https://github.com/K1rakishou/Kuroba-Experimental/
 *
 * This program is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with this program.  If not, see <http://www.gnu.org/licenses/>.
 */
package com.github.k1rakishou.chan;

import android.app.Activity;
import android.app.Application;
import android.content.Context;
import android.os.Build;
import android.os.Bundle;

import androidx.annotation.NonNull;

import com.github.k1rakishou.chan.core.cache.downloader.FileCacheException;
import com.github.k1rakishou.chan.core.di.AppModule;
import com.github.k1rakishou.chan.core.di.ExecutorsModule;
import com.github.k1rakishou.chan.core.di.GsonModule;
import com.github.k1rakishou.chan.core.di.HelperModule;
import com.github.k1rakishou.chan.core.di.LoaderModule;
import com.github.k1rakishou.chan.core.di.ManagerModule;
import com.github.k1rakishou.chan.core.di.NetModule;
import com.github.k1rakishou.chan.core.di.RepositoryModule;
import com.github.k1rakishou.chan.core.di.RoomDatabaseModule;
import com.github.k1rakishou.chan.core.di.SiteModule;
import com.github.k1rakishou.chan.core.di.UseCaseModule;
import com.github.k1rakishou.chan.core.manager.ApplicationVisibilityManager;
import com.github.k1rakishou.chan.core.manager.BoardManager;
import com.github.k1rakishou.chan.core.manager.ReportManager;
import com.github.k1rakishou.chan.core.manager.SettingsNotificationManager;
import com.github.k1rakishou.chan.core.manager.SiteManager;
import com.github.k1rakishou.chan.core.net.DnsSelector;
import com.github.k1rakishou.chan.core.settings.ChanSettings;
import com.github.k1rakishou.chan.ui.service.SavingNotification;
import com.github.k1rakishou.chan.ui.settings.SettingNotificationType;
import com.github.k1rakishou.chan.utils.AndroidUtils;
import com.github.k1rakishou.chan.utils.Logger;
import com.github.k1rakishou.common.AppConstants;
import com.github.k1rakishou.feather2.Feather;
import com.github.k1rakishou.model.DatabaseModuleInjector;
import com.github.k1rakishou.model.di.ModelMainComponent;

import org.greenrobot.eventbus.EventBus;
import org.jetbrains.annotations.NotNull;

import java.io.IOException;
import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import javax.inject.Inject;

import io.reactivex.exceptions.UndeliverableException;
import io.reactivex.plugins.RxJavaPlugins;
import kotlin.coroutines.CoroutineContext;
import kotlinx.coroutines.CoroutineScope;
import kotlinx.coroutines.Dispatchers;
import okhttp3.Dns;
import okhttp3.Protocol;

import static com.github.k1rakishou.chan.utils.AndroidUtils.getFlavorType;
import static com.github.k1rakishou.chan.utils.AndroidUtils.getVerifiedBuildType;
import static com.github.k1rakishou.chan.utils.AndroidUtils.postToEventBus;
import static java.lang.Thread.currentThread;

public class Chan
        extends Application
        implements Application.ActivityLifecycleCallbacks, CoroutineScope {
    private static final String TAG = "Chan";
    private int activityForegroundCounter = 0;

    @Inject
    SiteManager siteManager;
    @Inject
    BoardManager boardManager;
    @Inject
    ReportManager reportManager;
    @Inject
    SettingsNotificationManager settingsNotificationManager;
    @Inject
    ApplicationVisibilityManager applicationVisibilityManager;

    @NotNull
    @Override
    public CoroutineContext getCoroutineContext() {
        return Dispatchers.getMain();
    }

    private static Feather feather;

    /**
     * Only ever use this method in cases when there is no other way around it (like when normal
     * injection will create a circular dependency)
     * TODO(dependency-cycles): get rid of this method once all dependency cycles are resolved
     * */
    public static <T> T instance(Class<T> tClass) {
        return feather.instance(tClass);
    }

    public static <T> T inject(T instance) {
        feather.injectFields(instance);
        return instance;
    }

    @Override
    protected void attachBaseContext(Context base) {
        super.attachBaseContext(base);

        AndroidUtils.init(this);

        // spit out the build hash to the log
        AndroidUtils.getVerifiedBuildType();

        // remove this if you need to debug some sort of event bus issue
        EventBus.builder().logNoSubscriberMessages(false).installDefaultEventBus();
    }

    @Override
    public void onCreate() {
        super.onCreate();

        long start = System.currentTimeMillis();
        onCreateInternal();

        long diff = System.currentTimeMillis() - start;
        Logger.d(TAG, "Application initialization took " + diff + "ms");
    }

    private void onCreateInternal() {
        registerActivityLifecycleCallbacks(this);

        boolean isDev = getFlavorType() == AndroidUtils.FlavorType.Dev;
        boolean isBeta = getFlavorType() == AndroidUtils.FlavorType.Beta;

        System.setProperty(
                "kotlinx.coroutines.debug",
                isDev ? "on" : "off"
        );

        AppConstants appConstants = new AppConstants(getApplicationContext(), isDev);
        logAppConstants(appConstants);

        SavingNotification.setupChannel();

        Dns okHttpDns = getOkHttpDns();
        OkHttpProtocols okHttpProtocols = getOkHttpProtocols();

        ModelMainComponent modelMainComponent = DatabaseModuleInjector.build(
                this,
                okHttpDns,
                okHttpProtocols.protocols,
                Logger.TAG_PREFIX,
                isDev,
                isDev || isBeta,
                ChanSettings.verboseLogs.get(),
                appConstants,
                this
        );

        feather = Feather.with(
                new AppModule(this, this, okHttpDns, okHttpProtocols, appConstants),
                new UseCaseModule(),
                new ExecutorsModule(),
                // TODO: change to a normal dagger implementation when we get rid of Feather
                new RoomDatabaseModule(modelMainComponent),
                new NetModule(),
                new GsonModule(),
                new RepositoryModule(),
                new SiteModule(),
                new LoaderModule(),
                new ManagerModule(),
                new HelperModule()
        );
        feather.injectFields(this);

        siteManager.initialize();
        boardManager.initialize();

        setupErrorHandlers();

        // TODO(KurobaEx): move to background thread!
        if (ChanSettings.collectCrashLogs.get()) {
            if (reportManager.hasCrashLogs()) {
                settingsNotificationManager.notify(SettingNotificationType.CrashLog);
            }
        }
    }

    private void setupErrorHandlers() {
        RxJavaPlugins.setErrorHandler(e -> {
            if (e instanceof UndeliverableException) {
                e = e.getCause();
            }

            if (e == null) {
                return;
            }

            if (e instanceof IOException) {
                // fine, irrelevant network problem or API that throws on cancellation
                return;
            }
            if (e instanceof InterruptedException) {
                // fine, some blocking code was interrupted by a dispose call
                return;
            }
            if (e instanceof RuntimeException && e.getCause() instanceof InterruptedException) {
                // fine, DB synchronous call (via runTask) was interrupted when a reactive stream
                // was disposed of.
                return;
            }
            if (e instanceof FileCacheException.CancellationException
                    || e instanceof FileCacheException.FileNotFoundOnTheServerException) {
                // fine, sometimes they get through all the checks but it doesn't really matter
                return;
            }
            if ((e instanceof NullPointerException) || (e instanceof IllegalArgumentException)) {
                // that's likely a bug in the application
                currentThread().getUncaughtExceptionHandler().uncaughtException(currentThread(), e);
                return;
            }
            if (e instanceof IllegalStateException) {
                // that's a bug in RxJava or in a custom operator
                currentThread().getUncaughtExceptionHandler().uncaughtException(currentThread(), e);
                return;
            }

            Logger.e(TAG, "RxJava undeliverable exception", e);
            onUnhandledException(e, exceptionToString(true, e));

            // Do not exit the app here! Most of the time an exception that comes here is not a
            // fatal one. We only want to log and report them to analyze later. The app should be
            // able to continue running after that.
        });

        Thread.setDefaultUncaughtExceptionHandler((t, e) -> {
            // if there's any uncaught crash stuff, just dump them to the log and exit immediately
            Logger.e(TAG, "Unhandled exception", e);
            onUnhandledException(e, exceptionToString(false, e));
            System.exit(999);
        });
    }

    private void logAppConstants(AppConstants appConstants) {
        Logger.d(TAG, "maxPostsCountInPostsCache = " + appConstants.getMaxPostsCountInPostsCache());
        Logger.d(TAG, "maxAmountOfPostsInDatabase = " + appConstants.getMaxAmountOfPostsInDatabase());
        Logger.d(TAG, "maxAmountOfThreadsInDatabase = " + appConstants.getMaxAmountOfThreadsInDatabase());
    }

    private Dns getOkHttpDns() {
        if (ChanSettings.okHttpAllowIpv6.get()) {
            Logger.d(AppModule.DI_TAG, "Using DnsSelector.Mode.SYSTEM");
            return new DnsSelector(DnsSelector.Mode.SYSTEM);
        }

        Logger.d(AppModule.DI_TAG, "Using DnsSelector.Mode.IPV4_ONLY");
        return new DnsSelector(DnsSelector.Mode.IPV4_ONLY);
    }

    @NonNull
    private OkHttpProtocols getOkHttpProtocols() {
        if (ChanSettings.okHttpAllowHttp2.get()) {
            Logger.d(AppModule.DI_TAG, "Using HTTP_2 and HTTP_1_1");
            return new OkHttpProtocols(Arrays.asList(Protocol.HTTP_2, Protocol.HTTP_1_1));
        }

        Logger.d(AppModule.DI_TAG, "Using HTTP_1_1");
        return new OkHttpProtocols(Collections.singletonList(Protocol.HTTP_1_1));
    }

    private boolean isEmulator() {
        return Build.MODEL.contains("google_sdk")
                || Build.MODEL.contains("Emulator")
                || Build.MODEL.contains("Android SDK");
    }

    private String exceptionToString(boolean isCalledFromRxJavaHandler, Throwable e) {
        try (StringWriter sw = new StringWriter()) {
            try (PrintWriter pw = new PrintWriter(sw)) {
                e.printStackTrace(pw);
                String stackTrace = sw.toString();

                if (isCalledFromRxJavaHandler) {
                    return "Called from RxJava onError handler.\n" + stackTrace;
                }

                return "Called from unhandled exception handler.\n" + stackTrace;
            }
        } catch (IOException ex) {
            throw new RuntimeException("Error while trying to convert exception to string!", ex);
        }
    }

    private void onUnhandledException(Throwable exception, String errorText) {
        Logger.e("UNCAUGHT", errorText);
        Logger.e("UNCAUGHT", "------------------------------");
        Logger.e("UNCAUGHT", "END OF CURRENT RUNTIME MESSAGES");
        Logger.e("UNCAUGHT", "------------------------------");
        Logger.e("UNCAUGHT", "Android API Level: " + Build.VERSION.SDK_INT);
        Logger.e("UNCAUGHT", "App Version: " + BuildConfig.VERSION_NAME + "." + BuildConfig.BUILD_NUMBER);
        Logger.e("UNCAUGHT", "Development Build: " + getVerifiedBuildType().name());
        Logger.e("UNCAUGHT", "Phone Model: " + Build.MANUFACTURER + " " + Build.MODEL);

        // don't upload debug crashes
        if ("Debug crash".equals(exception.getMessage())) {
            return;
        }

        if (isEmulator()) {
            return;
        }

        if (ChanSettings.collectCrashLogs.get()) {
            reportManager.storeCrashLog(exception.getMessage(), errorText);
        }
    }

    private void activityEnteredForeground() {
        boolean lastForeground = getApplicationInForeground();

        activityForegroundCounter++;

        if (getApplicationInForeground() != lastForeground) {
            Logger.d(TAG, "^^^ App went foreground ^^^");

            if (applicationVisibilityManager != null) {
                applicationVisibilityManager.onEnteredForeground();
            }

            postToEventBus(new ForegroundChangedMessage(getApplicationInForeground()));
        }
    }

    private void activityEnteredBackground() {
        boolean lastForeground = getApplicationInForeground();

        activityForegroundCounter--;
        if (activityForegroundCounter < 0) {
            activityForegroundCounter = 0;
        }

        if (getApplicationInForeground() != lastForeground) {
            Logger.d(TAG, "vvv App went background vvv");

            if (applicationVisibilityManager != null) {
                applicationVisibilityManager.onEnteredBackground();
            }

            postToEventBus(new ForegroundChangedMessage(getApplicationInForeground()));
        }
    }

    public boolean getApplicationInForeground() {
        return activityForegroundCounter > 0;
    }

    public static class ForegroundChangedMessage {
        public boolean inForeground;

        public ForegroundChangedMessage(boolean inForeground) {
            this.inForeground = inForeground;
        }
    }

    //region Empty Methods
    @SuppressWarnings("EmptyMethod")
    @Override
    public void onActivityCreated(Activity activity, Bundle savedInstanceState) {
    }

    @Override
    public void onActivityStarted(Activity activity) {
        activityEnteredForeground();
    }

    @SuppressWarnings("EmptyMethod")
    @Override
    public void onActivityResumed(Activity activity) {
    }

    @SuppressWarnings("EmptyMethod")
    @Override
    public void onActivityPaused(Activity activity) {
    }

    @Override
    public void onActivityStopped(Activity activity) {
        activityEnteredBackground();
    }

    @SuppressWarnings("EmptyMethod")
    @Override
    public void onActivitySaveInstanceState(Activity activity, Bundle outState) {
    }

    @SuppressWarnings("EmptyMethod")
    @Override
    public void onActivityDestroyed(Activity activity) {
    }
    //endregion Empty Methods

    public static class OkHttpProtocols {
        public final List<Protocol> protocols;

        public OkHttpProtocols(List<Protocol> okHttpProtocols) {
            this.protocols = okHttpProtocols;
        }
    }
}

File: Kuroba/app/src/main/java/com/github/k1rakishou/chan/Chan.kt
package com.github.k1rakishou.chan

import android.app.Activity
import android.app.Application
import android.app.Application.ActivityLifecycleCallbacks
import android.content.Context
import android.os.Bundle
import android.os.StrictMode
import android.os.SystemClock
import com.github.k1rakishou.chan.core.AppDependenciesInitializer
import com.github.k1rakishou.chan.core.di.component.application.ApplicationComponent
import com.github.k1rakishou.chan.core.di.component.application.DaggerApplicationComponent
import com.github.k1rakishou.chan.core.di.module.application.AppModule
import com.github.k1rakishou.chan.core.di.module.application.CrossModuleBridge
import com.github.k1rakishou.chan.core.di.module.application.JsonParserModule
import com.github.k1rakishou.chan.core.di.module.application.LoaderModule
import com.github.k1rakishou.chan.core.di.module.application.ManagerModule
import com.github.k1rakishou.chan.core.di.module.application.NetModule
import com.github.k1rakishou.chan.core.di.module.application.ParserModule
import com.github.k1rakishou.chan.core.di.module.application.RepositoryModule
import com.github.k1rakishou.chan.core.di.module.application.SiteModule
import com.github.k1rakishou.chan.core.di.module.application.UseCaseModule
import com.github.k1rakishou.chan.core.helper.ImageLoaderFileManagerWrapper
import com.github.k1rakishou.chan.core.helper.ImageSaverFileManagerWrapper
import com.github.k1rakishou.chan.core.helper.ThreadDownloaderFileManagerWrapper
import com.github.k1rakishou.chan.core.helper.migration.app.ApplicationMigrationHelper
import com.github.k1rakishou.chan.core.manager.ApplicationVisibilityManager
import com.github.k1rakishou.chan.ui.activity.CrashReportActivity
import com.github.k1rakishou.chan.ui.adapter.PostsFilter
import com.github.k1rakishou.chan.utils.AppModuleAndroidUtils
import com.github.k1rakishou.chan.utils.AppModuleAndroidUtils.getDimen
import com.github.k1rakishou.chan.utils.CrashStorage
import com.github.k1rakishou.chan.utils.TimeUtils
import com.github.k1rakishou.common.AndroidUtils
import com.github.k1rakishou.common.AppConstants
import com.github.k1rakishou.common.dns.DnsOverHttpsSelector
import com.github.k1rakishou.common.dns.DnsOverHttpsSelectorFactory
import com.github.k1rakishou.common.dns.NormalDnsSelector
import com.github.k1rakishou.common.dns.NormalDnsSelectorFactory
import com.github.k1rakishou.core_logger.Logger
import com.github.k1rakishou.core_spannable.SpannableModuleInjector
import com.github.k1rakishou.core_themes.ThemesModuleInjector
import com.github.k1rakishou.fsaf.BadPathSymbolResolutionStrategy
import com.github.k1rakishou.fsaf.FileManager
import com.github.k1rakishou.fsaf.manager.base_directory.DirectoryManager
import com.github.k1rakishou.model.di.ModelComponentInjector
import com.github.k1rakishou.v2.ApplicationSettingsParameters
import com.github.k1rakishou.v2.KurobaSettings
import com.github.k1rakishou.v2.NonBackupableSettingsParameters
import com.github.k1rakishou.v2.database.KurobaSettingsDatabase
import com.github.k1rakishou.v2.di.KurobaSettingsComponentInjector
import dagger.Lazy
import kotlinx.coroutines.CoroutineExceptionHandler
import kotlinx.coroutines.CoroutineName
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.DEBUG_PROPERTY_NAME
import kotlinx.coroutines.DEBUG_PROPERTY_VALUE_OFF
import kotlinx.coroutines.DEBUG_PROPERTY_VALUE_ON
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.SupervisorJob
import kotlinx.coroutines.launch
import kotlinx.coroutines.runBlocking
import okhttp3.HttpUrl.Companion.toHttpUrl
import okhttp3.OkHttpClient
import okhttp3.dnsoverhttps.DnsOverHttps
import org.joda.time.Duration
import org.joda.time.format.PeriodFormatterBuilder
import java.net.InetAddress
import javax.inject.Inject
import kotlin.system.exitProcess

class Chan : Application(), ActivityLifecycleCallbacks {
  private var activityForegroundCounter = 0
  private var startTime = 0L

  // Delay job creation here because we need to first set the kotlinx.coroutines.DEBUG_PROPERTY_NAME
  private val job by lazy { SupervisorJob(null) }
  private lateinit var applicationScope: CoroutineScope

  private val kurobaSettingsDatabase by lazy {
    KurobaSettingsDatabase.buildDatabase(this@Chan)
  }

  private val kurobaSettings by lazy {
    KurobaSettings.create(
      kurobaSettingsDatabase = kurobaSettingsDatabase,
      applicationSettingsInfo = ApplicationSettingsParameters(
        applicationId = BuildConfig.APPLICATION_ID,
        isTablet = AppModuleAndroidUtils.isTablet,
        defaultFilterOrderName = PostsFilter.CatalogSortingOrder.BUMP.orderName,
        isDevBuild = AppModuleAndroidUtils.isDevBuild,
        isBetaBuild = AppModuleAndroidUtils.isBetaBuild,
        bookmarkGridViewInfo = ApplicationSettingsParameters.BookmarkGridViewInfo(
          getDimen(R.dimen.thread_grid_bookmark_view_default_width),
          getDimen(R.dimen.thread_grid_bookmark_view_min_width),
          getDimen(R.dimen.thread_grid_bookmark_view_max_width)
        )
      ),
      nonBackupableSettingsParameters = NonBackupableSettingsParameters(
        applicationMigrationVersion = ApplicationMigrationHelper.LATEST_VERSION
      )
    )
  }

  private val coroutineExceptionHandler = CoroutineExceptionHandler { _, exception ->
    Logger.e(TAG, "Coroutine unhandled exception exception", exception)
    onUnhandledException(exception)
  }

  private val tagPrefix by lazy { AndroidUtils.applicationLabel.toString() + " | " }

  @Inject
  lateinit var appDependenciesInitializer: AppDependenciesInitializer
  @Inject
  lateinit var applicationVisibilityManager: Lazy<ApplicationVisibilityManager>
  @Inject
  lateinit var appConstants: Lazy<AppConstants>
  @Inject
  lateinit var applicationMigrationHelper: ApplicationMigrationHelper

  class NormalDnsSelectorFactoryImpl(
    private val kurobaSettings: KurobaSettings
  ) : NormalDnsSelectorFactory {
    override fun createDnsSelector(okHttpClient: OkHttpClient): NormalDnsSelector {
      Logger.deps("NormalDnsSelectorFactory")

      if (kurobaSettings.application.okHttpAllowIpv6.readBlocking()) {
        Logger.d(Logger.DI_TAG, "Using DnsSelector.Mode.SYSTEM")
        return NormalDnsSelector(NormalDnsSelector.Mode.SYSTEM)
      }

      Logger.d(Logger.DI_TAG, "Using DnsSelector.Mode.IPV4_ONLY")
      return NormalDnsSelector(NormalDnsSelector.Mode.IPV4_ONLY)
    }
  }

  class DnsOverHttpsSelectorFactoryImpl(
    private val kurobaSettings: KurobaSettings
  ) : DnsOverHttpsSelectorFactory {
    override fun createDnsSelector(okHttpClient: OkHttpClient): DnsOverHttpsSelector {
      Logger.deps("DnsOverHttpsSelectorFactory")

      val selector = DnsOverHttps.Builder()
        .includeIPv6(kurobaSettings.application.okHttpAllowIpv6.readBlocking())
        .client(okHttpClient)
        .url("https://cloudflare-dns.com/dns-query".toHttpUrl())
        .bootstrapDnsHosts(
          listOf(
            InetAddress.getByName("162.159.36.1"),
            InetAddress.getByName("162.159.46.1"),
            InetAddress.getByName("1.1.1.1"),
            InetAddress.getByName("1.0.0.1"),
            InetAddress.getByName("162.159.132.53"),
            InetAddress.getByName("2606:4700:4700::1111"),
            InetAddress.getByName("2606:4700:4700::1001"),
            InetAddress.getByName("2606:4700:4700::0064"),
            InetAddress.getByName("2606:4700:4700::6400")
          )
        )
        .build()

      return DnsOverHttpsSelector(selector)
    }
  }

  val applicationInForeground: Boolean
    get() = activityForegroundCounter > 0
  val appRunningTime: Long
    get() {
      if (startTime == 0L) {
        return 0
      }

      return SystemClock.elapsedRealtime() - startTime
    }

  override fun attachBaseContext(base: Context) {
    super.attachBaseContext(base)

    val isDev = AppModuleAndroidUtils.isDevBuild
    System.setProperty(
      DEBUG_PROPERTY_NAME,
      if (isDev) DEBUG_PROPERTY_VALUE_ON else DEBUG_PROPERTY_VALUE_OFF
    )

    AndroidUtils.init(this)
    AppModuleAndroidUtils.init(this)

    Logger.init(
      prefix = tagPrefix,
      isDevBuild = AppModuleAndroidUtils.isDevBuild,
      verboseLogs = kurobaSettings.application.verboseLogs.readBlocking(),
      appContext = this
    )
  }

  override fun onCreate() {
    super.onCreate()

    startTime = SystemClock.elapsedRealtime()

    val start = System.currentTimeMillis()
    onCreateInternal()
    val diff = System.currentTimeMillis() - start

    Logger.d(TAG, "Application initialization took " + diff + "ms")
  }

  private fun onCreateInternal() {
    registerActivityLifecycleCallbacks(this)
    applicationScope = CoroutineScope(job + Dispatchers.Main + CoroutineName("Chan") + coroutineExceptionHandler)

    val isDev = AppModuleAndroidUtils.isDevBuild
    if (isDev && ENABLE_STRICT_MODE) {
      StrictMode.setThreadPolicy(
        StrictMode.ThreadPolicy.Builder()
          .detectAll()
          .penaltyLog()
          .penaltyFlashScreen()
          .build()
      )

      StrictMode.setVmPolicy(
        StrictMode.VmPolicy.Builder()
          .detectAll()
          .penaltyLog()
          .build()
      )
    }

    val kurobaExUserAgent = buildString {
      append(AndroidUtils.applicationLabel)
      append(" ")
      append(BuildConfig.VERSION_NAME)
    }

    val isLowRamDevice = kurobaSettings.application.isLowRamDeviceBlocking()

    val appConstants = AppConstants(
      context = applicationContext,
      isLowRamDevice = isLowRamDevice,
      kurobaExCustomUserAgent = kurobaExUserAgent,
      overrideUserAgent = { kurobaSettings.application.customUserAgent.readBlocking() },
      maxPostsInDatabaseSettingValue = 75000,
      maxThreadsInDatabaseSettingValue = 12500
    )

    // Preload user-agent on a background thread
    applicationScope.launch(Dispatchers.IO) { appConstants.userAgentMightBeOverridden }

    val fileManager = provideApplicationFileManager()
    val imageSaverFileManagerWrapper =  provideImageSaverFileManagerWrapper()
    val threadDownloaderFileManagerWrapper =  provideThreadDownloaderFileManagerWrapper()
    val imageLoaderFileManagerWrapper =  provideImageLoaderFileManagerWrapper()

    val themeEngine = ThemesModuleInjector.build(
      application = this,
      scope = applicationScope,
      fileManager = fileManager,
      kurobaSettings = kurobaSettings
    ).getThemeEngine()

    val kurobaSettingsComponent = KurobaSettingsComponentInjector.build(
      application = this
    )

    val modelComponent = ModelComponentInjector.build(
      application = this,
      scope = applicationScope,
      normalDnsSelectorFactory = NormalDnsSelectorFactoryImpl(kurobaSettings),
      dnsOverHttpsSelectorFactory = DnsOverHttpsSelectorFactoryImpl(kurobaSettings),
      verboseLogs = kurobaSettings.application.verboseLogs.readBlocking(),
      isDevFlavor = isDev,
      isLowRamDevice = isLowRamDevice,
      okHttpUseDnsOverHttps = kurobaSettings.application.okHttpUseDnsOverHttps.readBlocking(),
      appConstants = appConstants,
      kurobaSettings = kurobaSettings
    )

    // We need to start initializing ChanPostRepository first because it deletes old posts during
    // the initialization.
    modelComponent.getChanPostRepository().initialize()

    applicationComponent = DaggerApplicationComponent.builder()
      .modelModuleComponent(modelComponent)
      .kurobaSettingsModuleComponent(kurobaSettingsComponent)
      .application(this)
      .appContext(this)
      .kurobaSettings(kurobaSettings)
      .themeEngine(themeEngine)
      .fileManager(fileManager)
      .imageSaverFileManagerWrapper(imageSaverFileManagerWrapper)
      .threadDownloaderFileManagerWrapper(threadDownloaderFileManagerWrapper)
      .imageLoaderFileManagerWrapper(imageLoaderFileManagerWrapper)
      .applicationCoroutineScope(applicationScope)
      .normalDnsSelectorFactory(NormalDnsSelectorFactoryImpl(kurobaSettings))
      .dnsOverHttpsSelectorFactory(DnsOverHttpsSelectorFactoryImpl(kurobaSettings))
      .appConstants(appConstants)
      .appModule(AppModule())
      .crossModuleBridge(CrossModuleBridge())
      .gsonModule(JsonParserModule())
      .loaderModule(LoaderModule())
      .managerModule(ManagerModule())
      .netModule(NetModule())
      .repositoryModule(RepositoryModule())
      .siteModule(SiteModule())
      .parserModule(ParserModule())
      .useCaseModule(UseCaseModule())
      .build()
      .also { component -> component.inject(this) }

    runBlocking { themeEngine.initialize(this@Chan, TimeUtils.isHalloweenToday()) }
    SpannableModuleInjector.initialize(themeEngine)

    appDependenciesInitializer.init()
    setupErrorHandlers()

    applicationMigrationHelper.processMigrations(this)
  }

  private fun setupErrorHandlers() {
    Thread.setDefaultUncaughtExceptionHandler { thread, e ->
      // if there's any uncaught crash stuff, just dump them to the log and exit immediately
      Logger.e(TAG, "Unhandled exception in thread: ${thread.name}", e)
      onUnhandledException(e)
    }
  }

  private fun onUnhandledException(exception: Throwable) {
    try {
      CrashStorage.saveCrash(
        context = this,
        exception = exception,
        userAgent = appConstants.get().userAgentMightBeOverridden,
        appLifeTime = formatAppRunningTime()
      )

      CrashReportActivity.launch(this)
    } finally {
      android.os.Process.killProcess(android.os.Process.myPid())
      exitProcess(-1)
    }
  }

  private fun activityEnteredForeground() {
    val lastForeground = applicationInForeground
    activityForegroundCounter++

    if (applicationInForeground != lastForeground) {
      Logger.d(TAG, "^^^ App went foreground ^^^")

      applicationVisibilityManager.get().onEnteredForeground()
    }
  }

  private fun activityEnteredBackground() {
    val lastForeground = applicationInForeground
    activityForegroundCounter--

    if (activityForegroundCounter < 0) {
      activityForegroundCounter = 0
    }

    if (applicationInForeground != lastForeground) {
      Logger.d(TAG, "vvv App went background vvv")

      applicationVisibilityManager.get().onEnteredBackground()
    }
  }

  /**
   * This is the main instance of FileManager that is used by the most of the app.
   * */
  private fun provideApplicationFileManager(): FileManager {
    val directoryManager = DirectoryManager(this)

    return FileManager(
      appContext = this,
      badPathSymbolResolutionStrategy = BadPathSymbolResolutionStrategy.ReplaceBadSymbols,
      directoryManager = directoryManager
    )
  }

  /**
   * This is a separate copy of FileManager that exist for the sole purpose of only being used by
   * ImageSaver. That's because all public methods of FileManager are globally locked and the SAF
   * version of the FileManager is slow as fuck so when you download albums the rest of the app will
   * HANG because of the FileManager methods will be locked. To avoid this situation we use a
   * second, separate, instance of FileManager that will only be used in ImageSaver so the other file
   * manager that is used by the app is not getting locked while the user downloads something.
   * */
  private fun provideImageSaverFileManagerWrapper(): ImageSaverFileManagerWrapper {
    val directoryManager = DirectoryManager(this)

    val fileManager = FileManager(
      appContext = this,
      badPathSymbolResolutionStrategy = BadPathSymbolResolutionStrategy.ReplaceBadSymbols,
      directoryManager = directoryManager
    )

    return ImageSaverFileManagerWrapper(fileManager)
  }

  private fun provideThreadDownloaderFileManagerWrapper(): ThreadDownloaderFileManagerWrapper {
    val directoryManager = DirectoryManager(this)

    val fileManager = FileManager(
      appContext = this,
      badPathSymbolResolutionStrategy = BadPathSymbolResolutionStrategy.ReplaceBadSymbols,
      directoryManager = directoryManager
    )

    return ThreadDownloaderFileManagerWrapper(fileManager)
  }

  private fun provideImageLoaderFileManagerWrapper(): ImageLoaderFileManagerWrapper {
    val directoryManager = DirectoryManager(this)

    val fileManager = FileManager(
      appContext = this,
      badPathSymbolResolutionStrategy = BadPathSymbolResolutionStrategy.ReplaceBadSymbols,
      directoryManager = directoryManager
    )

    return ImageLoaderFileManagerWrapper(fileManager)
  }

  fun formatAppRunningTime(): String {
    val time = appRunningTime
    if (time <= 0) {
      return "Unknown (appContext=${this::class.java.simpleName}), time ms: $time"
    }

    return appRunningTimeFormatter.print(Duration.millis(time).toPeriod())
  }

  override fun onActivityCreated(activity: Activity, savedInstanceState: Bundle?) {}
  override fun onActivityStarted(activity: Activity) {
    activityEnteredForeground()
  }

  override fun onActivityResumed(activity: Activity) {}
  override fun onActivityPaused(activity: Activity) {}
  override fun onActivityStopped(activity: Activity) {
    activityEnteredBackground()
  }

  override fun onActivityDestroyed(activity: Activity) {}
  override fun onActivitySaveInstanceState(activity: Activity, outState: Bundle) {}

  companion object {
    private const val TAG = "Chan"
    private const val ENABLE_STRICT_MODE = false

    private val appRunningTimeFormatter = PeriodFormatterBuilder()
      .printZeroAlways()
      .minimumPrintedDigits(2)
      .appendHours()
      .appendSuffix(":")
      .appendMinutes()
      .appendSuffix(":")
      .appendSeconds()
      .appendSuffix(".")
      .appendMillis3Digit()
      .toFormatter()

    private lateinit var applicationComponent: ApplicationComponent

    @JvmStatic
    fun getComponent(): ApplicationComponent {
      return applicationComponent
    }
  }
}